### DISCUSSION TITLE
Can't call methods after `ref()`ing sources

### DISCUSSION THREAD

**Discussion by danbo** [savonet/liquidsoap#2535] [User Support]
When looking at manually processing requests and using the following examples

https://github.com/savonet/liquidsoap/discussions/2113#discussioncomment-2429071
https://github.com/savonet/liquidsoap/discussions/1903

if a = request.dynamic(...), the above work
if a = ref(some default source, ie blank())

There's an error that a.**fetch** or a.**add** do not exist - i read that methods can be lost:
https://github.com/savonet/liquidsoap/issues/1827

What is the correct way to create a reference-able dynamic request source? or anything for that matter where you can specify a type vs a concrete default that may be harder to construct?

**Comment by vitoyucepi** [comment]
@danbo I'm not sure if I understand you correct, but `blank()` has no method `add()` at all.
This means that only `request.dynamic()` could be used, unless you use preprocessor directives(I think so)
```
%ifdef output.alsa
print("We have support for ALSA.")
%else
print("We don't have support for ALSA.")
%endif
```

It's hard for me to imagine the way you could use `if a = ref(blank())`.
So please post some code for me to use it as a pivot.

**Reply by danbo** [reply]
If you use the examples above, ie from #2113 

```
queue_when_this_amount_is_left = 90.

r = request.dynamic.list(...)

def queue(remaining, _) =
  log("Queueing with #{remaining} seconds remaining")
  if not r.fetch() then
    # You might want to recurse and/or handle errors here:
    log("Fetching next query failed")
  end
end

r = source.on_end(delay=queue_when_this_amount_is_left , r, queue)
```

and actually define the request callback and want to do something with the dynamic source in the callback, you would have to presumably ref the r before defining the callback so it would know about r

```
r = ref(request.dynamic.list(... , getRequest))

def getRequest() =
  ...
  !r.someDynamicSourceMethodLikeAddOrFetch()
end
```

but then you can't do that because the ref refers to a callback that is defined after it is referenced.

You are right, sources such as blank don't have an add() or a seek(), but then what should I be ref'ing so that when I want to reference a ref'd r, it won't complain about missing add/fetch if request.dynamic returns a source (which is presumably more decorated than a normal source)?

This stemmed from wanting to control the queue, for which #2113 and #1903 show to achieve, but the moment you need to do something more specialized like r.seek() inside the callback and require r to be ref'd to access since it's not passed in, then all of a sudden there's an issue where fetch() and add() are no longer found. Remove the ref and it works again.

I'm trying to understand why code that might work for r, stops working when r = ref(...) and I'm also trying to understand what ... in ref(...) needs to be for it to work / how to reference the return type of request.dynamic.

btw, if I do use r=ref(blank()) the seek() method will be found and work in the callback, but of course add() and fetch() wont' be found so queue control fails as soon as r becomes a ref(blank()) value.

So it comes down to how do you ref complex types in general that have been enriched by core method calls.

**Reply by vitoyucepi** [reply]
## `r` is reference
I used this trick before.

```ruby
r = ref(request.dynamic(fun() -> null()))

def get_request()
  print((!r).add)
  null()
end

r := request.dynamic(get_request)
output.dummy(fallback([!r, blank()]))
```

1. `r = ref(request.dynamic(fun() -> null()))` - lets define a dummy `request.dynamic`. Another way is to use `r = ref(null())`, but it will then require `null.get` and other `null` checks everywhere.
1. And here's the function that will be used to generate next request.
   It will resolve reference to `r` and then return a `null` value.
   ```ruby
   def get_request()
     print((!r).add)
     null()
   end
   ```
1. `r := request.dynamic(get_request)` - redefine the `r` with the new function.
   I'm not sure if previous value of `r` will be actually destroyed.
1. `output.dummy(fallback([!r, blank()]))` - do some output for test purpose.

## function is reference
Alternatively you can define function as a reference like in https://github.com/savonet/liquidsoap/blob/49faf0e17e4a5c1ba9db8ff4bcd9fc03bf3e5f20/libs/ffmpeg.liq#L164-L188

```ruby
get_request = ref({null()})

r = request.dynamic(
  fun() -> begin
    fn = !get_request
    fn()
  end
)

get_request := fun() -> begin
  print(r.add)
  null()
end

output.dummy(fallback([r, blank()]))
```

1. `get_request = ref({null()})` - a reference to a function that will return `null`.
   Type compatible with `fun() -> request?`.
1. ```ruby
   r = request.dynamic(
     fun() -> begin
       fn = !get_request
       fn()
     end
   )
   ```
   This is required, because `request.dynamic(!get_request)` will be dereferenced into the wrong function instance.

BTW, I've found out that `request.dynamic.list` is deprecated.

**Reply by danbo** [reply]
After seeing you suggesting a similar solution to what I previously tried (request.dynamic(fun() -> null)), I decided to dig deeper...

Working (no reference):
```
def get_request() =
  null()
end
r = request.dynamic(get_request)
r = source.on_end(r, fun(_,_) -> ())
r = source.on_track(r, fun(_) -> ())
r = metadata.map(fun(m) -> [],update=false,strip=true,insert_missing=false,r)
output.dummy(fallback([r, blank()]))
```

Not working (r as reference): this value has no method 'fetch' on any call to source.on_end, source.on_track, metadata.map (you can comment out the others to see the error appears for any of the three method calls.
```
r = ref(request.dynamic(fun() -> null()))
def get_request() =
  null()
end
r := request.dynamic(get_request)
r := source.on_end(!r, fun(_,_) -> ())
r := source.on_track(!r, fun(_) -> ())
r := metadata.map(fun(m) -> [],update=false,strip=true,insert_missing=false,!r)
output.dummy(fallback([r, blank()]))
```
It looks like once r is reassigned through these methods, it has 'lost' its request.dynamic identity?

It had a fetch method when we started...

```
# r = ref(request.dynamic(fun() -> null()));;
r :
  ref(source(audio='A, video='B, midi='C)
      .{
        add : (request) -> bool,
        current : () -> request?,
        duration : () -> float,
        elapsed : () -> float,
        fallible : bool,
        fetch : () -> bool,
        id : () -> string,
        is_active : () -> bool,
        is_ready : () -> bool,
        is_up : () -> bool,
        last_metadata : () -> [string * string]?,
        log :
        {level : (() -> int?).{set : (int) -> unit}
        },
        on_get_ready : ((() -> unit)) -> unit,
        on_leave : ((() -> unit)) -> unit,
        on_metadata : ((([string * string]) -> unit)) -> unit,
        on_shutdown : ((() -> unit)) -> unit,
        on_track : ((([string * string]) -> unit)) -> unit,
        queue : () -> [request],
        remaining : () -> float,
        seek : (float) -> float,
        self_sync : () -> bool,
        set_queue : ([request]) -> unit,
        skip : () -> unit,
        time : () -> float
      }) =
  ref(<source>.{time = <fun>, fallible = true, skip = <fun>, seek = <fun>, is_active = <fun>, is_up = <fun>, log = {level = <fun>.{set = <fun>}}, self_sync = <fun>, duration = <fun>, elapsed = <fun>, remaining = <fun>, on_track = <fun>, on_leave = <fun>, on_shutdown = <fun>, on_get_ready = <fun>, on_metadata = <fun>, last_metadata = <fun>, is_ready = <fun>, id = <fun>, current = <fun>, set_queue = <fun>, add = <fun>, queue = <fun>, fetch = <fun>})
# def get_request() =
  null()
end
;;
get_request : () -> 'a? = <fun>
# r := request.dynamic(get_request);;
- : unit = ()
# r := source.on_end(!r, fun(_,_) -> ());;
At line 7, char 5-38:

Error 5: this value has no method `fetch`
```

PS: using docker image savonet/liquidsoap:v2.1.0

**Reply by vitoyucepi** [reply]
Here's your idea, but I purposefully split sources.

```ruby
r = ref(request.dynamic(fun() -> null()))

def get_request()
  request.create("test.mp3")
end

def custom_on_start()
  print('custom on start')
  if not (!r).add(get_request()) then
    log('Error while queueing initial request!')
  end
end

def custom_on_end(a, b)
  print('custom on end')
  print(a)
  print(b)
  if not (!r).add(get_request()) then
    log('Error while queueing next request')
  end
end

def custom_on_track(a)
  print('custom on track')
  print(a)
end

def custom_metadata_map(a)
  print('custom metadata map')
  print(a)
  [('a', 'b'), ('b', 'a')]
end

r := request.dynamic(prefetch=0, get_request)
(!r).on_track(custom_on_track)
custom_on_start()

t = source.on_end(delay=1., !r, custom_on_end)
t = metadata.map(strip=true, update=false, insert_missing=false, custom_metadata_map, t)
t.on_track(custom_on_track)

output.dummy(fallible = true, t)
```

Now about how it'll work

1. Function to produce request with sample music file, because it's necessary for testing `on_end` and `on_track`
   ```ruby
   def get_request()
     request.create("test.mp3")
   end
   ```
1. To use `on_end` we have to disable `prefetch`.
   ```ruby
   r := request.dynamic(prefetch=0, get_request)
   ```
1. Because of disabled `prefetch`, we have to bootstrap the `request` source manually
   ```ruby
   def custom_on_start()
     print('custom on start')
     if not (!r).add(get_request()) then
       log('Error while queueing initial request!')
     end
   end

   custom_on_start()
   ```
1. `source.on_end` is a standalone method(why?) with different definition, so I have to define it as another variable. Also we have to use the result in `output`, or else `on_end` will never be called.
   ```ruby
   def custom_on_end(a, b)
     print('custom on end')
     print(a)
     print(b)
     if not (!r).add(get_request()) then
       log('Error while queueing next request')
     end
   end

   t = source.on_end(delay=1., !r, custom_on_end)
   output.dummy(fallible = true, t)
   ```
1. `map.metadata` will create another source, but with custom metadata.
   ```ruby
   def custom_metadata_map(a)
     print('custom metadata map')
     print(a)
     [('a', 'b'), ('b', 'a')]
   end
   t = metadata.map(strip=true, update=false, insert_missing=false, custom_metadata_map, t)
   ```

I think this will work fine, but it's possible that requests will not be destroyed.

<details>
<summary>Maybe someday we will able to use this pieces of code.</summary>

```ruby
r = request.dynamic()
r.prefetch = 0
r.on_start(fun() -> ())
r.on_end(fun() -> ())
r.map_metadata(fun(_) -> [])
r.on_track(fun(_) -> ())
```
```ruby
r = request.dynamic()
r.prefetch = 1
r.resolve(fun() -> ())
```

</details>

**Reply by vitoyucepi** [reply]
But you can do even without refs if you want to `prefetch=0` and manually control `on_end`.
```ruby
r = request.dynamic(prefetch=0, {null()})

def custom_on_track(a)
  print('custom on track')
  print(a)
end

def get_request()
  print('get request')
  request.create("/var/liquidsoap/example.ogg")
end

def custom_on_start()
  print('custom on start')
  if not r.add(get_request()) then
    log('Error while queueing initial request!')
  end
end

def custom_on_end(a, b)
  print('custom on end')
  print(a)
  print(b)
  if not r.add(get_request()) then
    log('Error while queueing next request')
  end
end

def custom_metadata_map(a)
  print('custom metadata map')
  print(a)
  [('a', 'b'), ('b', 'a')]
end

r.on_track(custom_on_track)
custom_on_start()
t = source.on_end(delay=1., r, custom_on_end)
t = metadata.map(strip=true, update=false, insert_missing=false, custom_metadata_map, t)
t.on_track(custom_on_track)

output.dummy(fallible = true, t)
```

**Reply by rrolla** [reply]
@vitoyucepi i tried your example but getting error, maybe you have some idea where it need to nudge liquidsoap types?

here is my code
```rb
utilsUrl = getenv("UTILS_URL")
apiToken = getenv("API_TOKEN")

musicMainRequest = request.dynamic(id="musicMain", prefetch=0, {null()})

def api_call(channelId) =
    url = "#{utilsUrl}api/channels/#{channelId}/queue/next"
    print("API Sending GET request to '#{url}'")
    try
        response = http.get(url,
            headers=[
                ("Content-Type","application/json"),
                ("Accept","application/json"),
                ("Authorization","Bearer #{apiToken}")
            ],
        )
        print("API #{url} - Response status: #{response.status_code}")
        {success = response.status_code == 200, result = "#{response}"}
    catch err do
        print("API #{url} - Error: #{error.kind(err)} - #{error.message(err)}")
        {success = false, result = ""}
    end
end

def dynamic_request(channelId) =
    response = api_call(channelId)
    if (response.success != true) or (response.result == "") or (string.match(pattern="Error", response.result)) then
        null()
    else
        relativePath = response.result
        fullPath = "annotate:source=\"music\":/mnt/music/#{relativePath}"
        r = request.create(fullPath)
        if request.resolve(r) then
            r
        else
            null()
       end
    end
end

def main_custom_on_track(a)
  print('custom on track channel 1')
  print(a)
end

def main_get_request()
  print('get request channel 1')
  dynamic_request(1)
end

def main_custom_on_start()
  print('custom on start channel 1')
  if not musicMainRequest.add(main_get_request()) then
    print('Error while queueing initial request for channel 1!')
  end
end

def custom_on_end(a, b)
  print('custom on end channel 1')
  print(a)
  print(b)
  if not musicMainRequest.add(main_get_request()) then
    print('Error while queueing next request for channel 1!')
  end
end

musicMainRequest.on_track(main_custom_on_track)
main_custom_on_start()

musicMain = source.on_end(delay=1., musicMainRequest, main_custom_on_end)
musicMain.on_track(main_custom_on_track)

```

Error what i get
```rb
rolla-liquidsoap | At /home/liquidsoap/config/extras/api-fetch.liq, line 53, char 30-48:
rolla-liquidsoap |   if not musicMainRequest.add(main_get_request()) then
rolla-liquidsoap | 
rolla-liquidsoap | Error 5: this value has type
rolla-liquidsoap |   _? (inferred at /home/liquidsoap/config/extras/api-fetch.liq, line 26, char 4-34)
rolla-liquidsoap | but it should be a subtype of
rolla-liquidsoap |   request
rolla-liquidsoap exited with code 1

```

**Reply by toots** [reply]
`main_get_request` can return `null()` when no request can be produced, which needs to be handled. Something like this:
```ruby
def main_custom_on_start()
  print('custom on start channel 1')
  req = main_get_request()
  if not (null.defined(req) and musicMainRequest.add(null.get(req))) then
    print('Error while queueing initial request for channel 1!')
  end
end

def custom_on_end(a, b)
  print('custom on end channel 1')
  print(a)
  print(b)
  req = main_get_request()
  if not (null.defined(req) and musicMainRequest.add(null.get(req))) then
    print('Error while queueing next request for channel 1!')
  end
end
```
There could be other ways to implement this, for instance raising an error. However, using too many errors raising patterns can potentially lead to uncaught error which will crash a streaming script so handling via `null()` generally feels safer in this context.

**Reply by rrolla** [reply]
> `main_get_request` can return `null()` when no request can be produced, which needs to be handled. Something like this:
> 
> ```ruby
> def main_custom_on_start()
>   print('custom on start channel 1')
>   req = main_get_request()
>   if not (null.defined(req) and musicMainRequest.add(null.get(req))) then
>     print('Error while queueing initial request for channel 1!')
>   end
> end
> 
> def custom_on_end(a, b)
>   print('custom on end channel 1')
>   print(a)
>   print(b)
>   req = main_get_request()
>   if not (null.defined(req) and musicMainRequest.add(null.get(req))) then
>     print('Error while queueing next request for channel 1!')
>   end
> end
> ```
> 
> There could be other ways to implement this, for instance raising an error. However, using too many errors raising patterns can potentially lead to uncaught error which will crash a streaming script so handling via `null()` generally feels safer in this context.

Thanks for help @toots now working
